### PR TITLE
[MM][Perf][CG] Enable encoder CUDA Graph for MiniCPM-V

### PR DESCRIPTION
**Pull request by YunzhuLu**
## Purpose
Add encoder CUDA Graph support for MiniCPM-V 2.5, 2.6, 4.0  as part of tracker #38175. This implementation follows the existing workflow introduced in #38061. 

The captured graph covers both the ViT encoder (VPM) and the resampler.

MiniCPM-V 2.0 is not included, as it predates the slice-based vision architecture required by this implementation.
MiniCPM‑V 4.5 is not included, as its dynamic frame fusion introduces input-dependent encoder shapes that are not compatible with the current static-shape CUDA Graph capture mechanism. It continues to use the eager path and does not benefit from encoder CUDA Graph replay, which may reduce performance.

## Key Updates
This PR extends the multimodal encoder CUDA Graph cache to key captured graphs by` (token_budget, secondary_capture_axis_key)` instead of `token_budget `alone.

`token_budget` continues to determine greedy packing and output buffer sizing for resampler tokens. However, for vision encoders such as those in OpenBMB's MiniCPM-V models, the encoder capture shape also depends on discrete patch-grid / pixel-layout tiers. As a result, different input shapes can share the same token budget while requiring distinct CUDA Graph captures.

Without this secondary capture axis, inputs with the same token budget are forced into a single capture shape, which can introduce unnecessary padding and reduce the efficiency benefits of CUDA Graphs. By including `secondary_capture_axis_key `in the cache key, the encoder can maintain separate captures for shape-dependent tiers while preserving the existing token-budget-based scheduling and memory allocation logic.

## Test Plan
**Unit test**
```bash
pytest tests/v1/cudagraph/test_encoder_cudagraph.py -v
```

**Functional test**

GPU: RTX 5090   
Model: MiniCPM-V-4_0

No CUDA Graph
```bash
vllm serve /root/autodl-tmp/huggingface/hub/MiniCPM-V-4/OpenBMB/MiniCPM-V-4 \
  --trust-remote-code \
  --served-model-name MiniCPM-V-4 \
  --gpu-memory-utilization 0.75 \
  --max-model-len 4096 \
  --max-num-batched-tokens 4096 \
  --limit-mm-per-prompt '{"video": 1, "image": 1}'
```
With CUDA Graph
```bash
 vllm serve /root/autodl-tmp/huggingface/hub/MiniCPM-V-4/OpenBMB/MiniCPM-V-4 \
  --trust-remote-code \
  --served-model-name MiniCPM-V-4 \
  --gpu-memory-utilization 0.75 \
  --max-model-len 4096 \
  --max-num-batched-tokens 4096 \
  --limit-mm-per-prompt '{"video": 1, "image": 1}' \
  --compilation-config '{
    "cudagraph_mm_encoder": true,
    "encoder_cudagraph_token_budgets": [1024],
    "encoder_cudagraph_max_vision_items_per_batch": 32
  }'
```

**benchmark**

GPU: RTX 5090
Model: MiniCPM-V-2_6 / MiniCPM-V-4_0

```bash
vllm bench mm-processor \
  --model /root/autodl-tmp/huggingface/hub/MiniCPM-V-4/OpenBMB/MiniCPM-V-4 \
  --trust-remote-code \
  --tokenizer-mode slow \
  --dataset-name random-mm \
  --num-prompts 500 \
  --num-warmups 100 \
  --max-model-len 4096 \
  --seed 42 \
  --gpu-memory-utilization 0.8 \
  --random-mm-base-items-per-request 8 \
  --random-mm-num-mm-items-range-ratio 0.0 \
  --random-mm-bucket-config '{"(448, 448, 1)": 1.0}' \
  --compilation-config '{
    "cudagraph_mm_encoder": true, 
    "encoder_cudagraph_token_budgets": [256, 512], 
    "encoder_cudagraph_max_vision_items_per_batch": 8
  }'
```

## Test Result
✅ **Unit test**
```bash
36 passed, 16 warnings in 7.11s 
```

✅**Functional test**

Image
```bash
# No CUDA Graph
--------------------------------------------------
A woman is standing in a room with a television screen in the background displaying an image of a ship and some text in Chinese.
--------------------------------------------------
A woman is standing in a room with a television screen in the background displaying what appears to be a news broadcast about an aircraft carrier.
--------------------------------------------------
# With CUDA Graph 
--------------------------------------------------
A woman is standing in a room with a television screen in the background displaying an image of a ship and some text in Chinese.
--------------------------------------------------
A woman in a green t-shirt and denim shorts standing in a room with a television screen in the background displaying what appears to be a news broadcast about an aircraft carrier.
--------------------------------------------------
```

Video
```bash
# No CUDA Graph
--------------------------------------------------
The video captures a serene evening scene featuring a train crossing a bridge over a body of water. The sky is overcast with a mix of dark and light clouds, suggesting it might be either dawn or dusk. The train, illuminated by its own lights, moves steadily across the bridge, creating a bright line against the dimly lit sky. The reflection of the train's lights shimmers on the water's surface below, adding to the tranquil atmosphere. As the train progresses, the background reveals faint outlines of industrial structures, possibly factories or power plants, with their silhouettes barely visible against the horizon. The overall ambiance is calm and quiet, with the train's movement being the primary action in the scene.
--------------------------------------------------
The video captures a serene evening scene featuring a train crossing a bridge over a body of water. The sky is overcast with varying shades of gray, suggesting it might be either dawn or dusk. The train, illuminated by its own lights, moves steadily across the bridge, casting a reflection on the water below. As the train progresses, the lights on the train create a shimmering effect on the water's surface. In the background, faint outlines of industrial structures can be seen, adding an urban element to the otherwise natural landscape. The overall ambiance of the video is calm and tranquil, with the train's movement providing a subtle dynamic element to the scene.
--------------------------------------------------
# With CUDA Graph 
--------------------------------------------------
The video captures a serene evening scene featuring a train crossing a bridge over a body of water. The sky is overcast with a mix of dark and light clouds, suggesting it might be either dawn or dusk. The train, illuminated by its own lights, moves steadily across the bridge, creating a bright line against the dimly lit sky. The reflection of the train's lights shimmers on the water's surface below, adding to the tranquil atmosphere. As the train progresses, the background reveals faint outlines of industrial structures, possibly factories or power plants, with their silhouettes barely visible against the horizon. The overall ambiance is calm and quiet, with the train's movement being the primary action in the scene.
--------------------------------------------------
The video captures a serene evening scene featuring a train crossing a bridge over a body of water. The sky is overcast with varying shades of gray, suggesting it might be either dawn or dusk. The train, illuminated by its own lights, moves steadily across the bridge, casting a reflection on the water below. As the train progresses, the lights on the train create a shimmering effect on the water's surface. In the background, faint outlines of industrial structures can be seen, adding an urban element to the otherwise natural landscape. The overall ambiance of the video is calm and tranquil, with the train's movement providing a subtle dynamic element to the scene.
--------------------------------------------------
```

✅ **Benchmark:**

> Single RTX 5090 · random-mm · 500 prompts / 100 warmups

Benchmark results for MiniCPM-V 2.6 and 4.0 show no measurable regression. Latency remains within normal run-to-run variance, with small improvements observed in some runs.

| Version | Backend | Mean | P99 |
|------|------|------|------|
| MiniCPM-V-2_6 | FLASH_ATTN2 | +0.60% (44555.14→44286.37ms) | +0.10% (67712.77ms→67644.05ms) |
| MiniCPM-V-4 | FLASH_ATTN2 | +2.01% (44723.41→43822.66ms) | +2.69% (84580.66ms→82305.93ms) |

**MiniCPM-V-2_6**
No CUDA Graph:
```bash
================================================================================
Multimodal Processor Benchmark Results
================================================================================

MM Processor Metrics:
                     Stage  Mean Median   Std  P99.0
          get_mm_hashes_ms  0.35   0.35  0.03   0.39
get_cache_missing_items_ms  0.02   0.02  0.00   0.03
     apply_hf_processor_ms 30.16  29.88  1.43  35.51
        merge_mm_kwargs_ms  0.44   0.41  0.08   0.63
   apply_prompt_updates_ms  6.15   6.39  0.74   7.54
     preprocessor_total_ms 37.13  36.90  1.77  42.96
        encoder_forward_ms 64.14  63.93 14.47 111.22
         num_encoder_calls  1.09   1.00  0.28   2.00

Summary: 544 total encoder calls across 500 requests.

End-to-End Latency (ms):
Metric Value (ms)
  Mean   44723.41
Median   43812.82
   Std   13619.35
 P99.0   84580.66
```

With CUDA Graph:
```bash
================================================================================
Multimodal Processor Benchmark Results
================================================================================

MM Processor Metrics:
                     Stage  Mean Median   Std  P99.0
          get_mm_hashes_ms  0.35   0.35  0.02   0.38
get_cache_missing_items_ms  0.02   0.02  0.00   0.03
     apply_hf_processor_ms 30.12  29.93  1.31  34.74
        merge_mm_kwargs_ms  0.44   0.41  0.06   0.55
   apply_prompt_updates_ms  6.16   6.40  0.73   7.38
     preprocessor_total_ms 37.09  36.92  1.55  43.03
        encoder_forward_ms 62.67  61.63 14.23 113.54
         num_encoder_calls  1.09   1.00  0.28   2.00

Summary: 543 total encoder calls across 500 requests.

End-to-End Latency (ms):
Metric Value (ms)
  Mean   43822.66
Median   42867.49
   Std   13261.40
 P99.0   82305.93
```

**MiniCPM-V-4_0**
No CUDA Graph:
```bash
================================================================================
Multimodal Processor Benchmark Results
================================================================================

MM Processor Metrics:
                     Stage  Mean Median   Std  P99.0
          get_mm_hashes_ms  0.35   0.35  0.03   0.39
get_cache_missing_items_ms  0.02   0.02  0.00   0.03
     apply_hf_processor_ms 30.16  29.88  1.43  35.51
        merge_mm_kwargs_ms  0.44   0.41  0.08   0.63
   apply_prompt_updates_ms  6.15   6.39  0.74   7.54
     preprocessor_total_ms 37.13  36.90  1.77  42.96
        encoder_forward_ms 64.14  63.93 14.47 111.22
         num_encoder_calls  1.09   1.00  0.28   2.00

Summary: 544 total encoder calls across 500 requests.

End-to-End Latency (ms):
Metric Value (ms)
  Mean   44723.41
Median   43812.82
   Std   13619.35
 P99.0   84580.66
```

With CUDA Graph:
```bash
================================================================================
Multimodal Processor Benchmark Results
================================================================================

MM Processor Metrics:
                     Stage  Mean Median   Std  P99.0
          get_mm_hashes_ms  0.35   0.35  0.02   0.38
get_cache_missing_items_ms  0.02   0.02  0.00   0.03
     apply_hf_processor_ms 30.12  29.93  1.31  34.74
        merge_mm_kwargs_ms  0.44   0.41  0.06   0.55
   apply_prompt_updates_ms  6.16   6.40  0.73   7.38
     preprocessor_total_ms 37.09  36.92  1.55  43.03
        encoder_forward_ms 62.67  61.63 14.23 113.54
         num_encoder_calls  1.09   1.00  0.28   2.00

Summary: 543 total encoder calls across 500 requests.

End-to-End Latency (ms):
Metric Value (ms)
  Mean   43822.66
Median   42867.49
   Std   13261.40
 P99.0   82305.93
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/model_executor/models/minicpmv.py [hunk 1]
```diff
@@ -1219,6 +1225,392 @@ def get_vision_hidden_states(self, data: MiniCPMVImagePixelInputs) -> torch.Tens
...
+        col_idx = torch.arange(max_patches, device=device)
+        dummy_patch_mask = col_idx.unsqueeze(0) < dummy_patches_per_slice.unsqueeze(1)
+        buffers: dict[str, torch.Tensor] = {
+            _MINICPMV_CUDAGRAPH_BUF_KEY_TGT_SIZES: dummy_tgt_sizes,
+            _MINICPMV_CUDAGRAPH_BUF_KEY_PATCH_MASK: dummy_patch_mask,
+        }
+        if self.version == (4, 5):
+            buffers[_MINICPMV_CUDAGRAPH_BUF_KEY_TEMPORAL_IDS] = torch.zeros(
+                max_num_slices, dtype=torch.long, device=device
+            )
```
[the remark this answers is not in the document]
Comment (YunzhuLu) [reply vllm/model_executor/models/minicpmv.py]: fixed

File: vllm/model_executor/models/minicpmv.py [hunk 2]
```diff
@@ -1219,6 +1225,392 @@ def get_vision_hidden_states(self, data: MiniCPMVImagePixelInputs) -> torch.Tens
...
+            all_pixel_values,
+            patch_attention_mask=patch_attention_mask,
+            tgt_sizes=vpm_tgt_sizes,
+        )
+
+        if self.version == (4, 5):
+            temporal_ids = buffers[_MINICPMV_CUDAGRAPH_BUF_KEY_TEMPORAL_IDS]
+            resampler_out = self.resampler(vision_embedding, tgt_sizes, temporal_ids)
+        else:
+            resampler_out = self.resampler(vision_embedding, tgt_sizes)
```
[the remark this answers is not in the document]
Comment (YunzhuLu) [reply vllm/model_executor/models/minicpmv.py]: The resampler has been refactored so that all shape-dependent logic is now tensor-based.

File: vllm/model_executor/models/minicpmv.py [hunk 3]
```diff
@@ -1219,6 +1225,392 @@ def get_vision_hidden_states(self, data: MiniCPMVImagePixelInputs) -> torch.Tens
...
+        vpm_tgt_sizes = None if self.version == (2, 5) else tgt_sizes
+        vision_embedding = self.vpm(
+            all_pixel_values,
+            patch_attention_mask=patch_attention_mask,
+            tgt_sizes=vpm_tgt_sizes,
+        )
+
+        if self.version == (4, 5):
+            temporal_ids = buffers[_MINICPMV_CUDAGRAPH_BUF_KEY_TEMPORAL_IDS]
+            resampler_out = self.resampler(vision_embedding, tgt_sizes, temporal_ids)
```
[the remark this answers is not in the document]
Comment (YunzhuLu) [reply vllm/model_executor/models/minicpmv.py]: Resampler4_5 (MiniCPM-V4.5) is no longer supported in the CUDA Graph path in this PR

File: vllm/model_executor/models/minicpmv.py [hunk 4]
```diff
@@ -1219,6 +1225,392 @@ def get_vision_hidden_states(self, data: MiniCPMVImagePixelInputs) -> torch.Tens
...
+            temporal_ids = mm_kwargs.get("temporal_ids")
+            if temporal_ids is not None:
+                # temporal_ids is list[list[int]] (per-image, per-slice).
+                # After select_encoder_cudagraph_items it already holds only the
+                # selected items, so flatten directly to a 1-D tensor.
+                flat_ids = torch.tensor(
+                    flatten_2d_lists(temporal_ids), dtype=torch.long, device=device
+                )
+            else:
+                flat_ids = torch.zeros(len(tgt_sizes), dtype=torch.long, device=device)
```
[the remark this answers is not in the document]
Comment (YunzhuLu) [reply vllm/model_executor/models/minicpmv.py]: fixed
